### PR TITLE
Feature/cancellation reason ui

### PR DESCRIPTION
**Pull request by benRmc**
**Description**
Adds cancellation reason UI

**Issues that should be CLOSED by merge of this PR:**
* Fixes #

**Related issues that should be LINKED to this PR:**
* Connects #

### CODE REVIEW

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 1]
```diff
@@ -15,50 +15,127 @@ protocol CancellationViewControllerDelegate: class {
...
+        let required = (selectedReason?.explanationRequired ?? false) ? "(Required)" : "(Optional)"
+        explanationInputTextField.placeholder = "Please provide an explanation. \(required)"
+
+        tableView.reloadRows(at: [previousIndexPath, selectedReasonIndexPath].compactMap({$0}), with: .automatic)
     }
 
     @IBAction func didConfirm(_ sender: Any) {
-        if let quote = quote {
+        guard let quote = quote else {
+            Log("Cancellation quotet not found.", data: nil, level: .error)
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: Spelling: `... quote...`

File: traveler-swift-core/TravelerKit/AuthPath.swift [hunk 2]
```diff
@@ -132,7 +132,7 @@ enum AuthPath {
         case .cancelOrder(let request):
             urlComponents.path = "/v1/order/\(request.quote.order.id)/cancellation/\(request.quote.id)"
             urlRequest.jsonBody = [
-                "cancellationReason": request.reason,
+                "cancellationReason": request.reason.id,
```
Comment (OmarPadierna) [inline traveler-swift-core/TravelerKit/AuthPath.swift]: This does not belong in this PR. UI PRs should only contain UI layer changes. This is a core change.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 3]
```diff
@@ -15,50 +15,127 @@ protocol CancellationViewControllerDelegate: class {
...
+        if section == 0 {
+            return " "
+        } else {
+            return nil
+        }
     }
 
     override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        let cell = tableView.dequeueReusableCell(withIdentifier: "productCancelCell", for: indexPath) as! InfoCell
+        if indexPath.section == 0 {
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: favor switches instead of if statements for this kind of comparisons.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 4]
```diff
@@ -15,50 +15,127 @@ protocol CancellationViewControllerDelegate: class {
...
+            if indexPath.row >= (quote?.products.count ?? 0) {
+                let cell = tableView.dequeueReusableCell(withIdentifier: "productCancelSummaryCell", for: indexPath) as! InfoCell
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: "productCancelSummaryCell" is too similar to "productCancellCell". This makes it difficult to know what the cell represents at a first glance. I would rename "productCancellSummaryCell" to "totalRefundCell" since that is what that cell is actually representing.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 5]
```diff
@@ -15,50 +15,127 @@ protocol CancellationViewControllerDelegate: class {
...
+    }
+
+    // MARK: UITableViewDelegate
 
-        cell.titleLabel.text = quote?.products[indexPath.row].title
-        let percentageFee = quote!.percentageFee(product: quote!.products[indexPath.row])
-        cell.valueLabel.text = "Cancellation Fee \(percentageFee)%"
-        cell.secondValueLabel?.text = quote?.products[indexPath.row].cancellationCharge.localizedDescriptionInBaseCurrency
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard indexPath.section == 1 else {
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: 1) you can remove interactivity for individual cells in the storyboard so that you don't have to guard for this. 
2) This is another case when a switch is a better option. Running a switch on the section would achieve the same outcome you have with your guard statement with the advantage that if in the future we need to add more interactive sections to this table, you only need to add a case rather than rewriting this method.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 6]
```diff
@@ -15,50 +15,127 @@ protocol CancellationViewControllerDelegate: class {
...
 
     @IBAction func didConfirm(_ sender: Any) {
-        if let quote = quote {
+        guard let quote = quote else {
+            Log("Cancellation quotet not found.", data: nil, level: .error)
+            return
+        }
+
+        guard let reason = selectedReason else {
+            let alert = UIAlertController(title: "Cancellation reason not selected", message: nil, preferredStyle: .alert)
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: If a cancellation reason is required to move on with the flow, make the wording of the alerts imperative -> "Please select a cancellation reason"

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 7]
```diff
@@ -15,50 +15,127 @@ protocol CancellationViewControllerDelegate: class {
...
+            let okAction = UIAlertAction(title: "Ok", style: .cancel, handler: nil)
+            alert.addAction(okAction)
+            present(alert, animated: true, completion: nil)
+
+            return
+        }
+
+        do {
+            let cancellationRequest = try CancellationRequest(quote: quote, reason: reason, explanation: explanation)
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: This kind of error handling does not belong here. It belongs to the delegates. When initializing you should be able to do so without having to do error handling. If the initialization requires validation then you do an optional initializer. Will talk more in depth about this on the core PR.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 8]
```diff
@@ -68,6 +145,14 @@ class CancellationViewController: UITableViewController {
     }
 }
 
+extension CancellationViewController: UITextFieldDelegate {
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        view.endEditing(true)
+        explanation = (textField.text?.isEmpty ?? true) ? nil : textField.text
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: This line is not working properly. When writing an explanation and cancelling my request, the payload does not send the text in that field.

File: traveler-ios-ui/TravelerKitUI/ModelExtensions/CancellationError+LocalizedError.swift [hunk 9]
```diff
@@ -16,6 +16,8 @@ extension CancellationError: LocalizedError {
             return NSLocalizedString("Quote has expired", comment: "CancellationError")
         case .notCancellable:
             return NSLocalizedString("Order not cancellable", comment: "CancellationError")
+        case .explanationRequired:
+            return NSLocalizedString("An cancellation explanation is required", comment: "CancellationError")
```
[the remark this answers is not in the document]
Comment (benRmc) [reply traveler-ios-ui/TravelerKitUI/ModelExtensions/CancellationError+LocalizedError.swift]: This means a written explanation is required.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift (hunk 8 above)
Comment (benRmc) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: I have the explanation text in my JSON body. Could you please provide more details?

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift (hunk 7 above)
Comment (benRmc) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: It needs some discussion.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 10]
```diff
@@ -15,59 +15,134 @@ protocol CancellationViewControllerDelegate: class {
...
     override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        let cell = tableView.dequeueReusableCell(withIdentifier: "productCancelCell", for: indexPath) as! InfoCell
+        switch indexPath.section {
+        case 0:
+            if indexPath.row >= (quote?.products.count ?? 0) {
+                let cell = tableView.dequeueReusableCell(withIdentifier: "totalRefundCell", for: indexPath) as! InfoCell
+                cell.secondValueLabel?.text = quote?.totalRefund.localizedDescriptionInBaseCurrency
+
+                return cell
+            } else {
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: These are too nested..use tuples and pattern matching in your switch statement instead

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 11]
```diff
@@ -15,59 +15,134 @@ protocol CancellationViewControllerDelegate: class {
...
     }
 
     @IBAction func didClose(_ sender: Any) {
         dismiss(animated: true, completion: nil)
     }
 }
 
+extension CancellationViewController: UITextFieldDelegate {
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        view.endEditing(true)
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: You shouldn't have to become the delegate of the UITextField here

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 12]
```diff
@@ -15,59 +15,134 @@ protocol CancellationViewControllerDelegate: class {
...
+
+                cell.titleLabel.text = quote?.products[indexPath.row].title
+                let percentageFee = quote!.percentageFee(product: quote!.products[indexPath.row])
+                cell.valueLabel.text = "Cancellation Fee \(percentageFee)%"
+                cell.secondValueLabel?.text = quote?.products[indexPath.row].cancellationCharge.localizedDescriptionInBaseCurrency
+
+                return cell
+            }
+        default:
+            let cell = tableView.dequeueReusableCell(withIdentifier: "cancelReasonCell", for: indexPath) as! InfoCell
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: Never use default either...all your cases should be known

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 13]
```diff
@@ -15,59 +15,134 @@ protocol CancellationViewControllerDelegate: class {
...
 
+    private var selectedReason: CancellationReason?
+    private var explanation: String?
+    private var selectedReasonIndexPath: IndexPath?
+
     override func viewDidLoad() {
         super.viewDidLoad()
 
-        totalRefundLabel.text = quote?.totalRefund.localizedDescriptionInBaseCurrency
+        explanationInputTextField.delegate = self
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: You should use events instead of delegates here. The UITextFieldDelegate is used for something else

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 14]
```diff
@@ -15,57 +15,124 @@ protocol CancellationViewControllerDelegate: class {
...
 class CancellationViewController: UITableViewController {
-    @IBOutlet weak var totalRefundLabel: UILabel!
+    @IBOutlet weak var explanationInputTextField: UITextField!
 
     var quote: CancellationQuote?
     weak var delegate: CancellationViewControllerDelegate?
 
+    private var selectedReason: CancellationReason?
+    private var explanation: String?
+    private var selectedReasonIndexPath: IndexPath?
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: dont need both the indexpath and the reason

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 15]
```diff
@@ -15,57 +15,124 @@ protocol CancellationViewControllerDelegate: class {
...
+        if section == 0 {
+            return (quote?.products.count ?? 0) + 1
+        } else {
+            return (quote?.cancellationReasons.count ?? 0)
+        }
+    }
+
+    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        if section == 1 {
+            return "Please select a reason for cancellation"
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: why are we using titleheaders for this?

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 16]
```diff
@@ -15,57 +15,124 @@ protocol CancellationViewControllerDelegate: class {
...
+        } else {
+            return nil
+        }
+    }
+
+    override func tableView(_ tableView: UITableView, titleForFooterInSection section: Int) -> String? {
+        if section == 0 {
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: no need for this method altogether

File: traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift [hunk 17]
```diff
@@ -15,57 +15,124 @@ protocol CancellationViewControllerDelegate: class {
...
+            return " "
+        } else {
+            return nil
+        }
     }
 
     override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        let cell = tableView.dequeueReusableCell(withIdentifier: "productCancelCell", for: indexPath) as! InfoCell
+        switch (indexPath.section, indexPath.row == (quote?.products.count ?? 0)) {
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift]: use a more descriptive tuple...use the row itself as oppose to the boolean condition..this is difficult to read
